picasso/src/main/java/com/squareup/picasso3/ImageDecoderFactory.java

```diff
@@ -0,0 +1,28 @@
+package com.squareup.picasso3;
+
+import androidx.annotation.Nullable;
+import java.util.List;
+import okio.BufferedSource;
+
+final class ImageDecoderFactory {
+
+  final List<ImageDecoder> decoders;
+
+  ImageDecoderFactory(List<ImageDecoder> decoders) {
+    this.decoders = decoders;
+  }
+
+  /**
+   * Returns the first {@link ImageDecoder} that can handle the supplied <code>source</code>.
+   * @param source The source of the image data.
+   * @return The first ImageDecoder that can decode the source, or null.
+   */
+  @Nullable ImageDecoder getImageDecoderForSource(BufferedSource source) {
+    for (ImageDecoder decoder : decoders) {
+      if (decoder.canHandleSource(source)) {
+        return decoder;
+      }
+    }
+    return null;
```

Comment:
throw the ISE here?

Reply:
In the current implementation the ISE is thrown one level up, where the `RequestHandler` has the information to create a useful error message, which includes the URI or file path of the request.

We could throw one here, then catch it one level up and wrap it with a useful message, or just let the caller handle it.